streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1219,44 +1219,52 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
+                    latch <- Promise.make[Nothing, Unit]
+                    innerStream = Stream
+                      .managed(permits.withPermitManaged)
+                      .flatMap(_ => Stream.bracket(latch.succeed(()))(_ => UIO.unit))
+                      .flatMap(_ => f(a))
+                      .foreach(b => out.offer(Pull.emit(b)).unit)
+                      .foldCauseM(
+                        cause => out.offer(Pull.halt(cause)) *> innerFailure.fail(cause).unit,
+                        _ => ZIO.unit
+                      )
+                    _ <- innerStream.fork
+                    // Make sure that the current inner stream has actually succeeded in acquiring
+                    // a permit before continuing. Otherwise we could reach the end of the stream and
+                    // acquire the permits ourselves before the inners had a chance to start.
+                    _ <- latch.await
+                  } yield true
+              }
+
             }.foldCauseM(
-                cause => (interruptInners.succeed(()) *> out.offer(Pull.halt(cause))).unit.toManaged_,
+                cause => (ZIO.children.flatMap(Fiber.interruptAll) *> out.offer(Pull.halt(cause))).unit.toManaged_,
                 _ =>
                   innerFailure.await
-                  // Important to use `withPermits` here because the finalizer below may interrupt
+                  // Important to use `withPermits` here because the ZManaged#fork below may interrupt
                   // the driver, and we want the permits to be released in that case
                     .raceWith(permits.withPermits(n.toLong)(ZIO.unit))(
                       // One of the inner fibers failed. It already enqueued its failure, so we
-                      // signal the inner fibers to interrupt. The finalizer below will make sure
+                      // interrupt the inner fibers. The finalizer below will make sure
                       // that they actually end.
-                      leftDone =
-                        (_, permitAcquisition) => interruptInners.succeed(()) *> permitAcquisition.interrupt.unit,
+                      leftDone = (_, permitAcquisition) =>
+                        ZIO.children.flatMap(Fiber.interruptAll) *> permitAcquisition.interrupt.unit,
                       // All fibers completed successfully, so we signal that we're done.
                       rightDone = (_, failureAwait) => out.offer(Pull.end) *> failureAwait.interrupt.unit
                     )
                     .toManaged_
               )
               // This finalizer makes sure that in all cases, the driver stops spawning new streams
               // and the inner fibers are signalled to interrupt and actually exit.
-              .ensuringFirst(interruptInners.succeed(()) *> permits.withPermits(n.toLong)(ZIO.unit))
+              .ensuringFirst(
+                stopIteration.set(true) *> ZIO.children.flatMap(Fiber.interruptAll) &> permits
```

Comment:
this is where I get lost a bit: isn't there a risk that `ZIO.children` picks up the fibers created in `&>` and kills the whole thing?

Reply:
Hmm I did not consider this. Let me consult the implementation

Reply:
Ok, good catch, this is a pretty subtle pitfall. It's not that `children` here picks up the fibers forked from `&>` but rather that the entire finalizer is run on a different fiber from the one that spawned the inner fibers.

So it's going to actually require an effort to make this interrupt the fibers. I'm wondering though if I should actually be doing this manually though.

Reply:
Thanks Regis, this actually helped me simplify the combinators quite a bit. Will push the changes tomorrow.

Reply:
Ha, I'm thrilled this helped! Looking forward to see your simplification. Yes it feels like interrupting fibers when the parent is done is part of autosupervision.

Reply:
Ok, there it is. We still have to interrupt them if one of them failed or if the outer stream failed. I think we can't rely on supervision there because it has to be done eagerly.

Reply:
There's also a pitfall with accessing `ZIO.children` inside the continuation to `raceWith` - we have to capture those outside.

The scoping for `children` can be definitely confusing, as it is not lexical. Wonder if we can do something about that?

Reply:
Yeah I just saw your change, I see the subtlety there. Interruption on error makes definitely sense.